### PR TITLE
Z80 M-cycles + spec48 io contention

### PR DESCRIPTION
**Pull request by holub**
z80: machine-cycles during instruction execution
spec48: read/writes contention based on z80 changes

Current implementation doesn't adjust to ideal state possibly because of incorrect timings in z80 which need further investigation.

Some tests state (first - before, second - after):
tacts_before
tacts_after
... and fuse
tacts_fuse

ulatest3_before
ulatest3_after

fpga48_before
fpga48_aftter

fusetest_after
fusetest_after

timings48v1_before
timings48v1_after

some tests without baseline:

shock_after
ula48
aqua

### CODE REVIEW

**Comment by ICEknigh7** [comment]
These changes seem to fix bug report 8266: https://mametesters.org/view.php?id=8266
ys36

Almost fixes 8265: https://mametesters.org/view.php?id=8265
aquaplan

Doesn't seem to change 8264 : https://mametesters.org/view.php?id=8264
darkstar

**Comment by holub** [comment]
> These changes seem to fix bug report 8266: https://mametesters.org/view.php?id=8266
> ys36
> 
> Almost fixes 8265: https://mametesters.org/view.php?id=8265
> aquaplan
> 
> Doesn't seem to change 8264 : https://mametesters.org/view.php?id=8264
> darkstar
> 
> 

Nice! Timings still not perfect (same as original) but we'll keep improving as we now have room for that.

File: src/mame/drivers/spectrum.cpp [hunk 1]
```diff
@@ -306,16 +307,42 @@ uint8_t spectrum_state::pre_opcode_fetch_r(offs_t offset)
...
+		target: 57600 (12288)
+	*/
+
+	if (m_contention_pattern.empty() || offset < 0x4000 || (offset > 0x7fff /*128: && offset < 0xc000*/)
+		|| vpos < get_screen_area().top() || vpos > get_screen_area().bottom())
+		return;
+
+	attotime shift = m_maincpu->clocks_to_attotime(1);
+	attotime contentended_pass = shift + m_screen->scan_period() - m_screen->time_until_pos(vpos + 1, get_screen_area().left());
+	attotime contentended_left = shift + m_screen->scan_period() - m_screen->time_until_pos(vpos + 1, get_screen_area().right() + 1);
```
Comment (DopefishJustin) [inline src/mame/drivers/spectrum.cpp:338]: Should be "contended"

**Comment by holub** [comment]
Planning more improvements. Appreciate any thoughts and testing over z80 devices besides spectrum.

**Comment by jackson2k2** [comment]
You might need to examine if these changes break system1.cpp (8-bit SEGA arcade games from 1981-1988) which rely on setting cc cycle tables.
https://github.com/mamedev/mame/blob/master/src/mame/drivers/system1.cpp#L241

**Comment by holub** [comment]
> You might need to examine if these changes break system1.cpp (8-bit SEGA arcade games from 1983-1988) which rely on setting cc cycle tables.
> https://github.com/mamedev/mame/blob/master/src/mame/drivers/system1.cpp#L241

I see now it will. Untill custom table will be adjusted to new timings description. But I moving slowly to it. Amstard, msx, system1 are on the list

**Comment by holub** [comment]
couple more tests...

cpc_before
cpc_before
cpc_after
cpc_after

system1
system1

msx
msx

**Comment by angelosa** [comment]
> It's interesting to see a new take on Spectrum contentions.
> 
> Something I want to point out: I think stalling in the driver-side (adjust_icount) might not be precise enough, I could be wrong though!
> 
> best of luck to you on this endeavor!
> 
> * dink

`adjust_icount` applies to scaled clock rather than master/bus clock, which is inconvenient most of the time.

**Comment by dinkc64** [comment]
Curiously, have the Speccy test results improved for ulatest3 and fpga48all?  Could you also try fusetest & shockdemo? (attached)
[fusetest_shock.zip](https://github.com/mamedev/mame/files/8357689/fusetest_shock.zip)

best regards,
- dink

**Comment by holub** [comment]
> Curiously, have the Speccy test results improved for ulatest3 and fpga48all?  Could you also try fusetest & shockdemo? (attached)
> [fusetest_shock.zip](https://github.com/mamedev/mame/files/8357689/fusetest_shock.zip)
> 
> best regards,
> - dink

FYI I posted screenshots in the description

**Comment by dinkc64** [comment]
I was asking if the test has improved with recent iteration of code, because the screenshots in description still show incorrect handling of contentions w/Speccy.

**Comment by holub** [comment]
> I was asking if the test has improved with recent iteration of code, because the screenshots in description still show incorrect handling of contentions w/Speccy.

I have no intend to make timings correct with this PR. First we need to build foundation for that. Rest comming later.

**Comment by dinkc64** [comment]
Understood

**Comment by holub** [comment]
Made more progress locally. Will share shortly

**Comment by holub** [comment]
> Curiously, have the Speccy test results improved for ulatest3 and fpga48all? Could you also try fusetest & shockdemo? (attached) [fusetest_shock.zip](https://github.com/mamedev/mame/files/8357689/fusetest_shock.zip)
> 
> best regards,
> 
>     * dink

You can check requested results in PR description ;)

**Comment by dinkc64** [comment]
holub nice work, getting close to perfection :)

**Comment by redcode** [comment]
Could you also fix the Z80 core to make `ex (sp),hl`, `ex (sp),ix`, `ex (sp),iy` `push RR`, `call`, `rst` and interrupt instruction responses to write to memory in reverse order?

These operations must write to memory first the most significant byte and then the most significant one, which have an impact in the emulation of the ZX Spectrum memory contention.

Take a look at [Mostek Z80 Microcomputer Devices Technical Manual - MK3880 Central Processig Unit](http://zxe.io/downloads/documents/Mostek%20Z80%20Microcomputer%20Devices%20Technical%20Manual%20-%20MK3880%20Central%20Processig%20Unit%20(1979)(Mostek)(en).pdf), pp. 87, 88, 90, 91.

**Comment by jackson2k2** [comment]
Minor note: was likely hoping that these timing fixes would fix a long standing issue in emulated Moon Cresta (mooncrst) vs. a real Moon Cresta PCB:
https://www.youtube.com/watch?v=zvgvL2G1Mrk

But sadly they don't. I think the problem is a driver issue at this point...

**Comment by redcode** [comment]
For reference:

```

.- - | INT | - - - - - - - - - - - - - -.  ---		     ---
:	       VBLANK			:   | 8		      |
:- - - - - - - - - - - - - - - - -.- - -:  ---	       ---    |
:	Invisible Top Border	  :	:   | 8		|     |
.---------------------------------.	:  ---   ---	|     |
|	 Visible Top Border	  |	:   | 48  |	|     |
|----.-----------------------.----|	:  ---    |	|     |
|L   |			     |R	  |	:   |	  |	|     |
|e   |			     |i	  |  H	:   |	  |	|     |
|f  B|			     |g	 B|  B	:   |	  |	|     |
|t  o|	       Paper	     |h	 o|  L	:   | 192 | 296 | 304 | 312
|   r|			     |t	 r|  A	:   |	  |	|     |
|   d|			     |	 d|  N	:   |	  |	|     |
|   e|			     |	 e|  K	:   |	  |	|     |
|   r|			     |	 r|	:   |	  |	|     |
|----'-----------------------'----|	:  ---    |	|     |
|	   Bottom Border	  |	:   | 56  |	|     |
'---------------------------------'- - -'  ---   ---   ---   ---

|----|-----------------------|----|
  48		256	       48
|---------------------------------|
		352
```

**Comment by holub** [comment]
> Comparing screenshots while testing I've noticed that, in the latest merged PR, the resolution has been changed to 352x288 when it used to be 352x296 which, according to [this post](https://spectrumcomputing.co.uk/forums/viewtopic.php?p=41956#p41956), is actually the full raster image.
> 
> Shouldn't this be reverted, since the cropping is supposed to be handled by the views inside the Video options (as in the Neo-Geo)?
> 
> EDIT: The Russian Spectrum-based arcade games also seem to share that 352x296 output resolution, according to these: http://adb.arcadeitalia.net/lista_mame.php?game_sourcefile=photon2.cpp&arcade_only=0&current_version=0 https://bootleggames.fandom.com/wiki/Photon_IK02

There you go: https://github.com/mamedev/mame/pull/9600

**Comment by DavidHaywood** [comment]
any reason this was marked as draft?  I think it would be a nice set of changes to go in after the code freeze of this release, so that it gets the widest possible window for testing before the release after

**Comment by holub** [comment]
> any reason this was marked as draft?  I think it would be a nice set of changes to go in after the code freeze of this release, so that it gets the widest possible window for testing before the release after

Oh. I forgot to revert status back. At this point I'm pretty happy with results. I'm aware about couple missing peaces buy making everything absolutely perfect may take some time.

**Comment by cuavas** [comment]
Out of curiosity, can this easily cater for stuff like the Osborne-1 where the CPU can only access RAM once out of every four CPU clock cycles?

**Comment by holub** [comment]
> Out of curiosity, can this easily cater for stuff like the Osborne-1 where the CPU can only access RAM once out of every four CPU clock cycles?

Not sure I understand the question but idea is if driver doesn't setup nomreq_cb then cpu only access RAM for regular rop(), arg(). In this case nomreq_* just making dummy icount increment.

**Comment by cuavas** [comment]
So what would be the proper way to do it?  Check current cycle count and adjust icount enough to make it an even multiple of four?  Or is there more to it?

**Comment by holub** [comment]
> So what would be the proper way to do it?  Check current cycle count and adjust icount enough to make it an even multiple of four?  Or is there more to it?

I think I get it now. In base z80 rop takes 4 clocks, rm,wm - 3. In the driver you can simply create "contention" for rm and wm. So you going to have always 4:
m_maincpu->adjust_icount(-1);

**Comment by holub** [comment]
> So what would be the proper way to do it?  Check current cycle count and adjust icount enough to make it an even multiple of four?  Or is there more to it?

Or we can refactor MTM macro and allow to define it. In this case changing current 3 to 4 should cover your case

**Comment by ICEknigh7** [comment]
Also, this might be relevant to this PR, perhaps?
> MAME has a problem with `reti` / `retn`
> it recognizes several undocumented `retn`s as `reti`
> I don't know if anyone has tested the Z80 CTC to check whether it really recognizes more than one version of reti
> I have a compile-time option in my Z80 for that

```
    -DZ80_WITH_UNOFFICIAL_RETI=(YES|NO)
        Configure the ED5Dh, ED6Dh and ED7Dh undocumented instructions as "reti"
        instead of "retn".
        The default is NO.
```
image

**Comment by jackson2k2** [comment]
> Minor note: was likely hoping that some of these timing fixes would fix a long standing issue in emulated Moon Cresta (mooncrst) vs. a real Moon Cresta PCB: https://www.youtube.com/watch?v=zvgvL2G1Mrk
> 
> But sadly they don't. I think the problem is a driver issue at this point...

hap says that a possible issue might be that mooncrst initializes it's random seed from the Z80's R register, which while easy to comprehend on the surface, would be indeed "hard to emulate correctly..." Would be good to settle another time...

**Comment by redcode** [comment]
Another still not emulated (in MAME) undocumented behavior is this:
https://spectrumcomputing.co.uk/forums/viewtopic.php?t=6102

**Comment by cuavas** [comment]
Let’s just concentrate on what this PR is supposed to do.  Trying to drag in more pet issues is likely to complicate things.  I’m going to delete comments derailing this.

**Comment by holub** [comment]
> This is another unemulated undocumented behavior:
> https://spectrumcomputing.co.uk/forums/viewtopic.php?t=6102
> 
> (unemulated in MAME)

I have Q reg on the list but with very low priority :) then other stuff is finalized

**Comment by redcode** [comment]
This is not about Q, it's about the flags of the block instructions: LDxR, CPxR, INxR and OTxR.

**Comment by redcode** [comment]
> > This is another unemulated undocumented behavior:
> > https://spectrumcomputing.co.uk/forums/viewtopic.php?t=6102
> > (unemulated in MAME)
> 
> I have Q reg on the list but with very low priority :) then other stuff is finalized

This is not about Q, it's about the flags of the block instructions: LDxR, CPxR, INxR and OTxR.

**Comment by Tafoid** [comment]
@holub Thanks for the PR and the work involved.
Fortunately, I was able to quickly find the added Diagnostic rom/bios but most of the time this isn't the case.
I'm humbly suggest that if you have rom data to add to source in the future that you ensure the devteam has access to it by following the instructions [HERE](https://wiki.mamedev.org/index.php?title=Submitting_Source_Code).

**Comment by holub** [comment]
> Minor note: was likely hoping that some of these timing fixes would fix a long standing issue in emulated Moon Cresta (mooncrst) vs. a real Moon Cresta PCB: https://www.youtube.com/watch?v=zvgvL2G1Mrk
> 
> But sadly they don't. I think the problem is a driver issue at this point...

Hey, @jackson2k2 . Are you able to perform same validation on existing master and see if there is any difference with behavior?

**Pull request merged**